streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverTracker.scala

```diff
@@ -488,7 +493,28 @@ class ReceiverTracker(ssc: StreamingContext, skipReceiverLaunch: Boolean = false
           registerReceiver(streamId, typ, host, executorId, receiverEndpoint, context.senderAddress)
         context.reply(successful)
       case AddBlock(receivedBlockInfo) =>
-        context.reply(addBlock(receivedBlockInfo))
+        if (WriteAheadLogUtils.isBatchingEnabled(ssc.conf, isDriver = true)) {
+          val f = Future {
+            synchronized {
+              if (active) {
+                addBlock(receivedBlockInfo)
+              } else {
+                throw new IllegalStateException("Receiver Tracker Endpoint shutdown.")
+              }
+            }
+          }(walBatchingThreadPool)
+          f.onComplete { result =>
+            synchronized {
```

Comment:
Yeah. That's' a valid race condition. Its safer and simpler to reason about if `active` is made `@volatile`. You can removed the synchronized in line 507.